library/core/src/num/niche_types.rs

```diff
@@ -119,45 +108,32 @@ impl Default for Nanoseconds {
     }
 }
 
-define_valid_range_type! {
-    pub struct NonZeroU8Inner(u8 as u8 in 1..=0xff);
-    pub struct NonZeroU16Inner(u16 as u16 in 1..=0xff_ff);
-    pub struct NonZeroU32Inner(u32 as u32 in 1..=0xffff_ffff);
-    pub struct NonZeroU64Inner(u64 as u64 in 1..=0xffffffff_ffffffff);
-    pub struct NonZeroU128Inner(u128 as u128 in 1..=0xffffffffffffffff_ffffffffffffffff);
-
-    pub struct NonZeroI8Inner(i8 as u8 in 1..=0xff);
-    pub struct NonZeroI16Inner(i16 as u16 in 1..=0xff_ff);
-    pub struct NonZeroI32Inner(i32 as u32 in 1..=0xffff_ffff);
-    pub struct NonZeroI64Inner(i64 as u64 in 1..=0xffffffff_ffffffff);
-    pub struct NonZeroI128Inner(i128 as u128 in 1..=0xffffffffffffffff_ffffffffffffffff);
-}
-
-#[cfg(target_pointer_width = "16")]
-define_valid_range_type! {
-    pub struct UsizeNoHighBit(usize as usize in 0..=0x7fff);
-    pub struct NonZeroUsizeInner(usize as usize in 1..=0xffff);
-    pub struct NonZeroIsizeInner(isize as usize in 1..=0xffff);
-}
-#[cfg(target_pointer_width = "32")]
-define_valid_range_type! {
-    pub struct UsizeNoHighBit(usize as usize in 0..=0x7fff_ffff);
-    pub struct NonZeroUsizeInner(usize as usize in 1..=0xffff_ffff);
-    pub struct NonZeroIsizeInner(isize as usize in 1..=0xffff_ffff);
-}
-#[cfg(target_pointer_width = "64")]
-define_valid_range_type! {
-    pub struct UsizeNoHighBit(usize as usize in 0..=0x7fff_ffff_ffff_ffff);
-    pub struct NonZeroUsizeInner(usize as usize in 1..=0xffff_ffff_ffff_ffff);
-    pub struct NonZeroIsizeInner(isize as usize in 1..=0xffff_ffff_ffff_ffff);
-}
+const MAX_32_MINUS_1: u32 = u32::MAX - 1;
+const MAX_64_MINUS_1: u64 = u64::MAX - 1;
+const HALF_USIZE: usize = usize::MAX >> 1;
 
 define_valid_range_type! {
-    pub struct U32NotAllOnes(u32 as u32 in 0..=0xffff_fffe);
-    pub struct I32NotAllOnes(i32 as u32 in 0..=0xffff_fffe);
-
-    pub struct U64NotAllOnes(u64 as u64 in 0..=0xffff_ffff_ffff_fffe);
-    pub struct I64NotAllOnes(i64 as u64 in 0..=0xffff_ffff_ffff_fffe);
+    pub struct NonZeroU8Inner(u8 is 1.., != 0);
+    pub struct NonZeroU16Inner(u16 is 1.., != 0);
+    pub struct NonZeroU32Inner(u32 is 1.., != 0);
+    pub struct NonZeroU64Inner(u64 is 1.., != 0);
+    pub struct NonZeroU128Inner(u128 is 1.., != 0);
+
+    pub struct NonZeroI8Inner(i8 is ..0 | 1.., != 0);
+    pub struct NonZeroI16Inner(i16 is ..0 | 1.., != 0);
+    pub struct NonZeroI32Inner(i32 is ..0 | 1.., != 0);
+    pub struct NonZeroI64Inner(i64 is ..0 | 1.., != 0);
+    pub struct NonZeroI128Inner(i128 is ..0 | 1.., != 0);
+
+    pub struct UsizeNoHighBit(usize is 0..=HALF_USIZE, <= HALF_USIZE);
+    pub struct NonZeroUsizeInner(usize is 1.., != 0);
+    pub struct NonZeroIsizeInner(isize is ..0 | 1.., != 0);
+
+    pub struct U32NotAllOnes(u32 is 0..=MAX_32_MINUS_1, <= MAX_32_MINUS_1);
```

Comment:
rather than having the `MAX_32_MINUS_1` constant, could you use
`u32 is 0..u32::MAX, < u32::MAX`? similarly for the `u64` equivalent